### PR TITLE
JordanWigner macros

### PR DESCRIPTION
**Pull request by rreiner-hqs**
Macros for generating python implementations of the Jordan-Wigner transformation have been added to struqture-py-macros

### CODE REVIEW

File: struqture-py/python_tests/mappings_test.py
Comment (kbarkhqs) [inline struqture-py/python_tests/mappings_test.py:1]: There are no assertion statements to check that the conversion has happened correctly in these tests

File: struqture-py-macros/src/mappings.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                /// Transform the given fermionic object into a spin object using
+                /// the Jordan Wigner mapping.
+                pub fn jordan_wigner(&self) -> #output_wrapper_type {
+                    #output_wrapper_type {
+                        internal: self.internal.jordan_wigner()
+                    }
+                }
+            }
+        } else {
+            if struct_name == "FermionProduct" {
```
Comment (kbarkhqs) [inline struqture-py-macros/src/mappings.rs:129]: How come you use `==` here and `.contains` above? Should this be the same in both places? (also same comment for the function above)

**Review by kbarkhqs — CHANGES_REQUESTED** [review]
This is looking alright, but you'll need to add `pyo3` tests before we can merge it. Take a look, in the `struqture-py/tests` folder

File: struqture-py-macros/src/mappings.rs (hunk 1 above)
Comment (rreiner-hqs) [reply struqture-py-macros/src/mappings.rs:129]: The output type for all fermion `System` types can be built automatically (it is just the corresponding spin `System` type, see lines 115-117), while for the other spin types we have to specify it manually one by one.

File: struqture-py/python_tests/mappings_test.py
Comment (rreiner-hqs) [reply struqture-py/python_tests/mappings_test.py:1]: Do you think I should add assertions regarding the ouput types, or to check the whole outputs in detail?  The latter is being done already in the Rust tests

File: struqture-py/tests/spins/plus_minus_product.rs [hunk 2]
```diff
@@ -617,3 +620,34 @@ fn test_from_error() {
...
+            .unwrap()
+            .downcast::<PyTuple>()
+            .unwrap()
+            .get_item(0)
+            .unwrap();
+
+        let number_modes = usize::extract(fo.call_method0("number_modes").unwrap()).unwrap();
+        let number_spins =
+            usize::extract(pp.call_method0("current_number_spins").unwrap()).unwrap();
+        assert_eq!(number_modes, number_spins)
```
Comment (kbarkhqs) [inline struqture-py/tests/spins/plus_minus_product.rs:651]: Ah the pmp should also contain this method but I seem to have forgotten it, I'll pm you

**Review by kbarkhqs — APPROVED** [review]

**Pull request merged**